### PR TITLE
Add an abstraction layer for registries [Draft]

### PR DESCRIPTION
**Pull request by Matyrobbrt**
This Draft PR aims to make registering easier and creating registry instances. It is still a WIP, but my goal is to wrap forge's deferred registers in order to use them with builders for easily instantiating `RegistryObject`s

### CODE REVIEW

**Comment by Matyrobbrt** [comment]
So, my idea would be to have something like this (using blocks as an example)
```java
public static final RegistryObject<TestBlock> TEST = BLOCKS.register("test", TestBlock::new)
            .harvestLevel(HarvestLevel.DIAMOND).harvestTool(HarvestTool.PICKAXE).dropSelf()
            .blockItem(item -> item.tab(CreativeModeTab.TAB_BREWING).lang(" test item yes")).build();
```
In this case `Block::new` represents a factory which takes in `Properties` and returns `Block`

The above code would:
- register the block with the provided properties 
- register the blockitem with the properties provided
- add the harvest tool, level and other tags to a static map which would generate the tags when the runData task is ran
- do anything else we feel like it should do, like generating a model for the block as well etc

**Comment by Affehund** [comment]
Definitely a very good idea. That can save us a lot of work (especially for tags and models, maybe implement basic block states for vanilla types like fences / stairs etc…) once it is finished. 
In my opinion it would be great if it won’t be too complicated to register and not too abstract (still having a simple system), but it seems good at the moment. 
Maybe a little documentation such as above for all the features of the new registry would be helpful.
Looking forward to see how it will go!

**Comment by Affehund** [comment]
Another thing that came in my mind where having a option for simple recipes (like ingots to block, or smelting recipes), but I don’t if that would be too much ☺️

**Comment by Matyrobbrt** [comment]
> In my opinion it would be great if it won’t be too complicated to register and not too abstract (still having a simple system), but it seems good at the moment.

That is my plan. I do want to keep it so that you can register a simple block with like 4 methods as you do normally, but I also want it to be able to do most things so that you can register both complex and simple blocks. As per the abstraction, the only abstract thing will be the builders and the wrapper around the Forge registry which is something that will only be used internally so you shouldn't need to know / care about the abstraction layer behind it. 

> Maybe a little documentation such as above for all the features of the new registry would be helpful

I definitely plan on documenting it

**Comment by ConnorTron110** [comment]
Could also be used to make the BlockItem and add that to the items registry.

**Comment by Matyrobbrt** [comment]
> Could also be used to make the BlockItem and add that to the items registry.

Yes, as I demonstrated in the first example, the Blockbuilder will have methods for the blockitem as well

**Comment by Matyrobbrt** [comment]
The Blocks and Items are done. I think that the PR is ready to be merged, as any other wrappers (for effects, BEs etc.) will be implemented as we go. I'd appreciate if multiple people would review it, and if you have any other suggestions for other methods / features please let me know. I will make this PR a normal one now.

File: src/main/java/io/github/darealturtywurty/ancientology/common/blocks/TestBlock.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package io.github.darealturtywurty.ancientology.common.blocks;
+
+import net.minecraft.world.level.block.Block;
+
+public class TestBlock extends Block {
```
Comment (ConnorTron110) [inline src/main/java/io/github/darealturtywurty/ancientology/common/blocks/TestBlock.java]: I Would say remove the test block and leave the commented out example in the Block Init Class

File: src/generated/resources/.cache/cache [hunk 2]
```diff
@@ -0,0 +1,4 @@
+3dccc00bd1235b8950b30e4285a1a9db0ae17b12 assets/ancientology/lang/en_us.json
+87091ae6f7115ad2627b2ed09c69eca31ea4cb3e data/ancientology/loot_tables/blocks/test.json
+dd585fcc15f55cc62e8d4639ae2bb8be1e3b6c5f data/minecraft/tags/blocks/mineable/pickaxe.json
+dd585fcc15f55cc62e8d4639ae2bb8be1e3b6c5f data/minecraft/tags/blocks/needs_diamond_tool.json
```
Comment (ConnorTron110) [inline src/generated/resources/.cache/cache]: Iirc we don't need this file so it could be added to gitignore

**Review by ConnorTron110 — COMMENTED** [review]
The two comments I have where more of a clean-up during your testing phase, other than that LGTM

File: src/generated/resources/.cache/cache (hunk 2 above)
Comment (Matyrobbrt) [reply src/generated/resources/.cache/cache]: We do actually, the datagen needs to have all the previous generated files cached in order to work

Comment (ConnorTron110) [reply src/generated/resources/.cache/cache]: What's its purpose?

File: src/main/java/io/github/darealturtywurty/ancientology/core/util/interfaces/package-info.java [hunk 3]
```diff
@@ -0,0 +1 @@
+package io.github.darealturtywurty.ancientology.core.util.interfaces;
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/ancientology/core/util/interfaces/package-info.java]: This should be removed.

File: src/main/java/io/github/darealturtywurty/ancientology/mixins/accessors/LanguageProviderAccessor.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+import net.minecraftforge.common.data.LanguageProvider;
+
+@Mixin(LanguageProvider.class)
+public interface LanguageProviderAccessor {
+
+    @Accessor
+    String getLocale();
+
+}
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/ancientology/mixins/accessors/LanguageProviderAccessor.java:14]: This should be replaced with an AT.

File: src/main/java/io/github/darealturtywurty/ancientology/mixins/accessors/package-info.java [hunk 5]
```diff
@@ -0,0 +1 @@
+package io.github.darealturtywurty.ancientology.mixins.accessors;
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/ancientology/mixins/accessors/package-info.java]: This should be removed.

File: src/main/java/io/github/darealturtywurty/ancientology/mixins/package-info.java [hunk 6]
```diff
@@ -0,0 +1 @@
+package io.github.darealturtywurty.ancientology.mixins;
```
Comment (DaRealTurtyWurty) [inline src/main/java/io/github/darealturtywurty/ancientology/mixins/package-info.java]: This should be removed.

File: src/main/resources/ancientology.mixins.json [hunk 7]
```diff
@@ -4,7 +4,7 @@
     "compatibilityLevel": "JAVA_17",
     "refmap": "ancientology.refmap.json",
     "mixins": [
-        
+        "accessors.LanguageProviderAccessor"
```
Comment (DaRealTurtyWurty) [inline src/main/resources/ancientology.mixins.json:7]: Replaced with AT.

**Review by DaRealTurtyWurty — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/darealturtywurty/ancientology/mixins/accessors/LanguageProviderAccessor.java (hunk 4 above)
Comment (Matyrobbrt) [reply src/main/java/io/github/darealturtywurty/ancientology/mixins/accessors/LanguageProviderAccessor.java:14]: You can't, it's a forge class

**Review by DaRealTurtyWurty — APPROVED** [review]

**Pull request merged**